### ISSUE TITLE
No ROUTINE in WHERE Clause

### ISSUE THREAD

**Issue by dominickdm** [Riverside-Software/sonar-openedge#554]
Is it possible to introduce a rule to detect routines used in WHERE clauses, the same way as the -noroutineinwhere parameter works? 

Restricts the invocation of the following from WHERE clauses:
* Class method
* User-defined function
* Class property, which implements the GET method using ABL statements

**Comment by gquerret** [comment]
Yes, can be done

**Comment by jurjendijkstra** [comment]
In our codebase we use functions a lot in where-clauses, but thes are always functions that do not have the problem. For example: each booking where booking.date > projectstartdate(selectedproject) 

I understand the intention of the rule (or the intention of the -noroutineinwhere parameter) , that things go reaaly wrong when the return-value of the routine would depend on the record, in my example if the return-value would depend on the "booking" record. We can not use this rule or parameter, because it makes no exception for functions that we know are safe to use in a where clause. But now we have the risk that someone accidently uses a different routine that is actually unsafe. 

So, the rule would be really great if it could somehow distinguish between safe and unsafe routines. Perhaps just by checking the routine name with a list of known safe routines?

**Comment by jurjendijkstra** [comment]
Hmmm it seems I had that same issue in 2006:    http://www.oehive.org/prolint/rules/where-udf

**Comment by gquerret** [comment]
First beta version of the rule included in develop branch, with:
 * Class method
 * User-defined function
Detection of properties with GET methods will come in second time.
List of excluded methods/calls can be provided as a parameter in the rule

Feedback requested as soon as the rule is available in master :-)

**Comment by Ton-Pe** [comment]
I noticed that simple `FIND Table WHERE Table.Field = TestMethod()` is reported correctly. But if the where-clause is `FIND Table WHERE Table.Field = THIS-OBJECT:TestMethod()` or ` FIND Table WHERE Table.Field = KernelHelper:Kernel:SysCtrlChar("Parameter", "Type")` these are not reported.

Anyway, this is important rule I think. I just investigated a bug that was caused by using method call in where-clause.

And just noticed that Gilles has commented "Detection of properties with GET methods will come in second time" that partly might relate to my second example :)
